Source/WebCore/dom/Element.cpp

```diff
@@ -4740,6 +4743,48 @@ bool Element::isSpellCheckingEnabled() const
     return true;
 }
 
+bool Element::isWritingSuggestionsEnabled() const
+{
+    // If none of the following conditions are true, then return `false`.
+
+    // `element` is an `input` element whose `type` attribute is in either the
+    // `Text`, `Search`, `URL`, `Email` state and is `mutable`.
+    auto isEligibleInputElement = [&] {
+        RefPtr input = dynamicDowncast<HTMLInputElement>(*this);
+        if (!input)
+            return false;
+
+        return !input->isDisabledFormControl() && input->supportsWritingSuggestions();
+    }();
+
+    // `element` is a `textarea` element that is `mutable`.
+    auto isTextArea = is<HTMLTextAreaElement>(*this);
```

Comment:
This also needs the disabled check. Not sure about the editing host stuff below.